landlab/components/species_evolution/zone_controller.py

```diff
@@ -0,0 +1,416 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""ZoneController of SpeciesEvolver."""
+import numpy as np
+from scipy.ndimage.measurements import label
+
+from . import ZoneTaxon, Zone
+from .record import Record
+from .zone import _update_zones
+
+
+class ZoneController(object):
+    """Controls zones and populates them with taxa.
+
+    This object manages 'zones' that are used to evaluate the spatial aspect
+    of taxa. A zone represents a portion of a model grid. It is made up of
+    spatially continuous grid nodes.
+
+    This controller creates zones using the parameter, ``zone_function`` set at
+    initialization. This function identifies all of the grid nodes where zones
+    are to be created. A zone is created for each cluster of spatially
+    continuous nodes. Zones are updated also using this function when the
+    ``run_one_step`` method of this controller is called.
+
+    The structure of an example model grid is diagrammed below to demonstrate
+    how zones are created. The grid contains six columns and six rows. In this
+    example, the function returns True where node values are greater than 0.
+    Nodes marked with an ``*`` are nodes where a zone exists and the mask is
+    True. All other nodes are marked with a ``·``. A zone is created for each
+    cluster of continuous nodes where the mask is True.
+
+    values         mask returned
+    evaluated      by zone function
+    0 0 0 0 0 0    · · · · · ·
+    0 0 0 5 4 2    · · · * * *
+    0 6 0 4 6 0    · * · * * ·
+    0 2 0 0 0 0    · * · · · ·
+    0 0 4 0 4 0    · · * · * ·
+    0 0 0 0 0 0    · · · · · ·
+
+    The above example is continued in the following four diagrams that
+    demonstrate how individual zones are identified. Each zone is marked with a
+    ``x``, ``o``, ``+``, or ``@``. Clusters can be identified using ``D8``
+    where diagonal neighbors are included or ``D4`` where diagonal neighbors
+    are excluded. A minimum zone area can be enforced with the ``minimum_area``
+    parameter.
+
+    D8             D4             D8             D4
+    min area = 0   min area = 0   min area = 2   min area = 2
+    · · · · · ·    · · · · · ·    · · · · · ·    · · · · · ·
+    · · · + + +    · · · + + +    · · · + + +    · · · + + +
+    · x · + + ·    · x · + + ·    · x · + + ·    · x · + + ·
+    · x · · · ·    · x · · · ·    · x · · · ·    · x · · · ·
+    · · x · o ·    · · @ · o ·    · · x · · ·    · · · · · ·
+    · · · · · ·    · · · · · ·    · · · · · ·    · · · · · ·
+
+    The grid boundary affects zone creation. Zones can include nodes along the
+    boundary, although because boundary nodes are not associated with cells
+    that have area, boundary nodes do not contribute to the area summation of
+    clusters. The area summation only takes into account the cells associated
+    with core nodes.
```

Comment:
I would be careful saying "nodes along the boundary". That could imply nodes along the perimeter, and in Landlab, those are always boundary nodes. But boundary nodes can also be in locations that are not the perimeter of the grid. So maybe reword to:

The grid perimeter affects zone creation. All perimeter nodes are boundary nodes in Landlab. Zones can include nodes along the perimeter, although because perimeter nodes are not associated with cells that have area, perimeter nodes do not contribute to the area summation of clusters. [NOW THIS IS ME GUESSING, BUT I THINK HAS TO BE RIGHT GIVEN THAT LAST SENTENCE] Any nodes that are designated as boundary nodes but do not lie on the perimeter of the grid are not included in the area summation. The area summation only takes into account the cells associated with core nodes.

Reply:
Also, given the discussion about perimeter/boundary, I would not include the 2 value in row five, column 6.

One more thing - given the discussion of D4 vs. D8, does this mean that a RasterModelGrid must be used?

Reply:
Modified text given your first comment.

Removing the 2 value in the upper grid example maybe requires larger grid examples to demonstrate the key combinations of D4, D8 and min area. The node of the 2 as a part of that zone doesn't contradict perimeter considerations as well as how they are handled by ZoneController.

ZoneController is limited to RasterModelGrid. SpeciesEvolver is not. Changed as I described in a previous response.